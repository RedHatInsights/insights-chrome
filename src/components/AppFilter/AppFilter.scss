@import '~@patternfly/patternfly/sass-utilities/_all';

.chr-c-page__app-filter-dropdown-menu {
  position: absolute;
  .pf-c-dropdown__menu::before {
    z-index: -1000;
  }
  .pf-c-dropdown__menu {
    bottom:0;
    left: 0;
    top: 70px;
    overflow-y:scroll;
    overflow-x:hidden;
    position: fixed; // allows menu to go full width
    .chr-app-filter-content {
      display: flex;
      flex-direction: column;
      align-self: stretch;
      .chr-l-flex-app-filter-search {
        padding: var(--pf-global--spacer--xl) 0 var(--pf-global--spacer--xl) 0;
        justify-content: center;
        .pf-c-text-input-group {
          width: 300px;
          .pf-c-search-input {
            display: flex;
            &__icon {
                color: var(--pf-global--BackgroundColor--dark-100);
            }
            &__text-input{
                color: var(--pf-global--BackgroundColor--dark-100);
            }
          }
        }
      }
      div.gallery {
        column-gap: var(--pf-global--spacer--lg);
        margin: 0 auto; // centers columns
      }
      div.galleryItem {
        display: block;
        padding-bottom: var(--pf-global--spacer--xl);
        width: 350px;
        overflow: hidden; // prevents padding from spilling into next column
        break-inside: avoid-column; // prevents padding from spilling into next column
        .pf-l-split {
          margin-top: auto;
          .pf-l-split__item {
            h4 {
              color: var(--pf-global--BackgroundColor--dark-100);
              font: "RedHatDisplay", "Overpass", overpass, helvetica, arial, sans-serif; //--pf-global--FontFamily--heading--sans-serif
              font-weight:  var(--pf-global--FontWeight--normal);
              margin-bottom: var(--pf-global--spacer--md);
              white-space: nowrap;
            }
            p {
              margin-bottom: var(--pf-global--spacer--sm);
            }
            a {
              white-space: nowrap;
              color: var(--pf-global--active-color--100);
            }
          }
        }
      }
    }
  }
  @media screen and (max-width: $pf-global--breakpoint--md) { 
    .pf-c-dropdown__menu {
      bottom:0; // allows scrolling
      top: 119px;
      div.gallery {
        column-count: 1;
      }
    }
  }
  @media screen and (min-width: 769px) and (max-width: 1099px) {
    .pf-c-dropdown__menu {
      bottom: 0;  // allows scrolling
      div.gallery {
        column-count: 2;
        width: calc( 350px * 2 );
      }
    }
  }
  @media screen and (min-width: 1100px) and (max-width: 1549px) { 
    div.gallery {
      column-count: 3;
      width: calc( 350px * 3 );
    }
  }
  @media screen and (min-width: 1550px) { 
    div.gallery {
      column-count: 4;
      width: calc( 350px * 4 );
    }
  }
  .pf-c-empty-state {
    h4 {
      color: var(--pf-global--BackgroundColor--dark-100);
      font-size: var(--pf-c-title--m-xl--FontSize);
    }
    button {
      color: var(--pf-global--active-color--100);
    }
  }
}
