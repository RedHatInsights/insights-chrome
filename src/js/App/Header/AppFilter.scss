@media screen and (max-width:768px) {
  .pf-c-page__header {
    &-tools-group {
      margin-left: 0 !important;
    }
    &-tools {
      padding-right: 0;
    }
  }
}

.pf-c-page__header {
  .pf-c-backdrop {
    z-index: -10;
  }
}

.ins-c-page__app-filter-dropdown {
  .pf-c-dropdown__toggle {
    display: inline-block;
    margin-left: 72px;
    margin-right: var(--pf-global--spacer--sm);
    position: relative;
    white-space: nowrap;
    &::before {
      --pf-c-dropdown__toggle--before--BorderTopColor: var(--pf-global--disabled-color--100);
      --pf-c-dropdown__toggle--before--BorderRightColor: var(--pf-global--disabled-color--100);
      --pf-c-dropdown__toggle--before--BorderLeftColor: var(--pf-global--disabled-color--100);
    }
  }
  @media screen and (max-width:768px) { 
      .pf-c-dropdown__toggle {
        margin-left: 10px;
      }
  }
  .pf-c-dropdown__menu::before {
    z-index: -1;
  }


  .pf-c-dropdown__menu {
    left: 0;
    top: 78px;
    overflow-y:scroll;
    overflow-x:hidden;
    position: fixed; // allows menu to go full width
    .content {
      display: flex;
      flex-direction: column;
      align-self: stretch;
      .search {
        padding: var(--pf-global--spacer--xl) 0 var(--pf-global--spacer--xl) 0;
        justify-content: center;
        .pf-c-search-input {
          display: flex;
          width: 300px;
          &__icon {
              color: var(--pf-global--BackgroundColor--dark-100);
          }
          &__text-input{
              color: var(--pf-global--BackgroundColor--dark-100);
          }
        }
      }
      div.gallery {
        column-gap: var(--pf-global--spacer--lg);
        margin: 0 auto; // centers columns
      }
      div.galleryItem {
        display: block;
        padding-bottom: var(--pf-global--spacer--xl);
        width: 350px;
        overflow: hidden; // prevents padding from spilling into next column
        break-inside: avoid-column; // prevents padding from spilling into next column
        .pf-l-split {
          margin-top: auto;
          .pf-l-split__item {
            h4 {
              color: var(--pf-global--BackgroundColor--dark-100);
              font: "RedHatDisplay", "Overpass", overpass, helvetica, arial, sans-serif; //--pf-global--FontFamily--heading--sans-serif
              font-weight:  var(--pf-global--FontWeight--normal);
              margin-bottom: var(--pf-global--spacer--md);
              white-space: nowrap;
            }
            p {
              margin-bottom: var(--pf-global--spacer--sm);
            }
            a {
              white-space: nowrap;
              color: var(--pf-global--active-color--100);
            }
          }
        }
      }
    }
  }
  @media screen and (max-width:768px) { 
    .pf-c-dropdown__menu {
      bottom:0; // allows scrolling
      div.gallery {
        column-count: 1;
      }
    }
  }
  @media screen and (min-width: 769px) and (max-width: 1099px) {
    .pf-c-dropdown__menu {
      bottom: 0;  // allows scrolling
      div.gallery {
        column-count: 2;
        width: calc( 350px * 2 );
      }
    }
  }
  @media screen and (min-width: 1100px) and (max-width: 1549px) { 
    div.gallery {
      column-count: 3;
      width: calc( 350px * 3 );
    }
  }

  @media screen and (min-width: 1550px) { 
    div.gallery {
      column-count: 4;
      width: calc( 350px * 4 );
    }
  }

  .pf-c-empty-state {
    --pf-c-empty-state__body--Color: #6a6e73;
    h4 {
      color: var(--pf-global--BackgroundColor--dark-100);
      font-size: var(--pf-c-title--m-xl--FontSize);
    }
    button {
      color: var(--pf-global--active-color--100);
    }
  }
}
