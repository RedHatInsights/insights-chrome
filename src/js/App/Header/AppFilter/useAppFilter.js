import axios from 'axios';
import { useEffect, useState } from 'react';
import { useSelector } from 'react-redux';
import { isBeta, getEnv } from '../../../utils';
import { evaluateVisibility } from '../../../utils/isNavItemVisible';

export const requiredBundles = ['application-services', 'openshift', 'insights', ...(getEnv() !== 'prod' ? ['edge'] : []), 'ansible', 'settings'];
const bundlesOrder = ['application-services', 'openshift', 'rhel', 'edge', 'ansible', 'settings', 'cost-management', 'subscriptions'];

function getBundleLink({ title, isExternal, href, routes, expandable, ...rest }) {
  const costLinks = [];
  const subscriptionsLinks = [];
  let url = href;
  let appId = rest.appId;
  if (expandable) {
    routes.forEach(({ href, title, ...rest }) => {
      if (href.includes('/openshift/cost-management') && rest.filterable !== false) {
        costLinks.push({ ...rest, href, title });
      }

      if (rest.filterable !== false && (href.includes('/insights/subscriptions') || href.includes('/openshift/subscriptions'))) {
        subscriptionsLinks.push({
          ...rest,
          href,
          title,
        });
      }

      if (!url && href.match(/^\//)) {
        url = isExternal ? href : href.split('/').slice(0, 3).join('/');
        appId = rest.appId ? rest.appId : appId;
      }
    });
  }

  return {
    ...rest,
    appId,
    isExternal,
    title,
    href: url,
    costLinks,
    subscriptionsLinks,
  };
}

const isDuplicate = (items, href) => !!items.find((item) => item.href === href);

const useAppFilter = () => {
  const isBetaEnv = isBeta();
  const [state, setState] = useState({
    isLoaded: false,
    isLoading: false,
    isOpen: false,
    filterValue: '',
    data: {
      'cost-management': {
        id: 'cost-management',
        title: 'Cost Management',
        links: [],
      },
      subscriptions: {
        id: 'subscriptions',
        title: 'Subscriptions',
        links: [],
      },
    },
  });
  const existingSchemas = useSelector(({ chrome: { navigation } }) => navigation);

  const handleBundleData = async ({ data: { id, navItems, title } }) => {
    let links = navItems
      .reduce((acc, curr) => {
        if (curr.groupId) {
          return [...acc, ...curr.navItems?.map(({ groupId, navItems, ...rest }) => (groupId ? navItems : rest))];
        }
        return [...acc, curr];
      }, [])
      .flat()
      .map(getBundleLink)
      .filter(({ filterable }) => filterable !== false);
    const bundleLinks = [];
    const extraLinks = {
      cost: [],
      subs: [],
    };
    const promises = links.map(async ({ costLinks, subscriptionsLinks, ...rest }) => {
      if (costLinks.length > 0) {
        extraLinks.cost = await costLinks.filter(evaluateVisibility);
      }

      if (subscriptionsLinks.length > 0) {
        extraLinks.subs = await subscriptionsLinks.filter(evaluateVisibility);
      }
      if (subscriptionsLinks.length > 0 || costLinks.length > 0) {
        return;
      }
      const link = await evaluateVisibility(rest);

      bundleLinks.push(link);
    });
    await Promise.all(promises);

    setState((prev) => ({
      ...prev,
      isLoaded: true,
      data: {
        ...prev.data,
        [id]: {
          id,
          title,
          links: bundleLinks,
        },
        'cost-management': {
          ...prev.data['cost-management'],
          links: [
            ...prev.data['cost-management'].links,
            ...extraLinks.cost.filter((item) => !isDuplicate(prev.data['cost-management'].links, item.href)),
          ],
        },
        subscriptions: {
          ...prev.data.subscriptions,
          links: [...prev.data.subscriptions.links, ...extraLinks.subs.filter((item) => !isDuplicate(prev.data.subscriptions.links, item.href))],
        },
      },
    }));
  };

  useEffect(() => {
    if (state.isOpen && !state.isLoading && !state.isLoaded) {
      setState((prev) => ({
        ...prev,
        isLoading: true,
      }));
      let bundles = requiredBundles.filter((app) => !Object.keys(existingSchemas).includes(app));
      bundles.map((fragment) =>
        axios
          .get(`${isBetaEnv ? '/beta' : ''}/config/chrome/${fragment}-navigation.json`)
          .then(handleBundleData)
          .then(() => Object.values(existingSchemas).map((data) => handleBundleData({ data })))
          .catch((err) => {
            console.error('Unable to load appfilter bundle', err);
          })
      );
    }
  }, [state.isOpen]);

  const setIsOpen = (isOpen) => {
    setState((prev) => ({
      ...prev,
      isOpen,
    }));
  };
  const setFilterValue = (filterValue = '') => {
    setState((prev) => ({
      ...prev,
      filterValue,
    }));
  };

  const filterApps = (data, filterValue = '') =>
    Object.entries(data).reduce(
      (acc, [key, { links, ...rest }]) => ({
        ...acc,
        [key]: {
          ...rest,
          links: links.filter(({ title, isHidden }) => !isHidden && title.toLowerCase().includes(filterValue.toLowerCase())),
        },
      }),
      {}
    );

  const filteredApps = filterApps(state.data, state.filterValue);
  return {
    ...state,
    setIsOpen,
    setFilterValue,
    filteredApps: bundlesOrder
      .map((app) => filteredApps[app])
      .filter((app) => typeof app !== 'undefined')
      .filter(({ links }) => links.length > 0),
  };
};

export default useAppFilter;
