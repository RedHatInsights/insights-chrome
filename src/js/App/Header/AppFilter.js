import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Button } from '@patternfly/react-core/dist/js/components/Button/Button';
import { CaretDownIcon } from '@patternfly/react-icons/dist/js/icons/caret-down-icon';
import { Dropdown } from '@patternfly/react-core/dist/js/components/Dropdown/Dropdown';
import { DropdownToggle } from '@patternfly/react-core/dist/js/components/Dropdown/DropdownToggle';
import { EmptyState, EmptyStateVariant } from '@patternfly/react-core/dist/js/components/EmptyState/EmptyState';
import { EmptyStateIcon } from '@patternfly/react-core/dist/js/components/EmptyState/EmptyStateIcon';
import { EmptyStateBody } from '@patternfly/react-core/dist/js/components/EmptyState/EmptyStateBody';
import FilterIcon from '@patternfly/react-icons/dist/js/icons/filter-icon';
import { Flex } from '@patternfly/react-core/dist/js/layouts/Flex/Flex';
import { SearchInput } from '@patternfly/react-core/dist/js/components/SearchInput/SearchInput';
import { Split } from '@patternfly/react-core/dist/js/layouts/Split/Split';
import { SplitItem } from '@patternfly/react-core/dist/js/layouts/Split/SplitItem';
import { TextContent } from '@patternfly/react-core/dist/js/components/Text/TextContent';
import { Text } from '@patternfly/react-core/dist/js/components/Text/Text';
import { Title } from '@patternfly/react-core/dist/js/components/Title/Title';
import { Spinner } from '@patternfly/react-core/dist/js/components/Spinner/Spinner';
import { Bullseye } from '@patternfly/react-core/dist/js/layouts/Bullseye/Bullseye';

import './AppFilter.scss';
import useGlobalNav from '../../utils/useGlobalNav';

const App = ({ id, title, routes, parent }) => (
  <div className="galleryItem">
    <Split>
      <SplitItem>
        <TextContent>
          <Text component="h4">{title}</Text>
          {routes.map((subApp) => {
            const redirectUrl = `${parent ? `${parent.id}/` : ''}${subApp.reload || `${id}/${subApp.id}`}`;

            return (
              <Text component="p" key={`${id}/${subApp.id}`}>
                <Text component="a" href={redirectUrl}>
                  {subApp.title}
                </Text>
              </Text>
            );
          })}
        </TextContent>
      </SplitItem>
    </Split>
  </div>
);

App.propTypes = {
  id: PropTypes.string,
  title: PropTypes.node,
  routes: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string.isRequired })),
  parent: PropTypes.shape({
    id: PropTypes.string,
  }),
};

const AppFilterDropdown = ({ isLoaded, setIsOpen, isOpen, filterValue, setFilterValue, filteredApps }) => (
  <Dropdown
    className="ins-c-page__app-filter-dropdown"
    onSelect={() => setIsOpen(true)}
    toggle={
      <DropdownToggle id="toggle-id" onToggle={() => setIsOpen(!isOpen)} toggleIndicator={CaretDownIcon}>
        Find an app or service
      </DropdownToggle>
    }
    isOpen={isOpen}
    ouiaId="App Filter"
  >
    <div className="content">
      {isLoaded ? (
        <React.Fragment>
          <Flex className="search">
            <SearchInput
              data-ouia-component-id="app-filter-search"
              placeholder="Find application or service"
              value={filterValue}
              onChange={(val) => setFilterValue(val)}
              onClear={() => setFilterValue('')}
            />
          </Flex>
          {filteredApps?.length > 0 ? (
            <div className="gallery">
              {filteredApps.map((app) => (
                <App key={app.id} {...app} />
              ))}
            </div>
          ) : (
            <EmptyState className="pf-u-mt-xl" variant={EmptyStateVariant.full}>
              <EmptyStateIcon className="pf-u-mb-xl" icon={FilterIcon} />
              <Title headingLevel="h4">No matching applications or services found.</Title>
              <EmptyStateBody className="pf-u-mb-xl">
                This filter criteria matches no applications or services. Try changing your input filter.
              </EmptyStateBody>
              <Button ouiaId="app-filter-clear-input" className="pf-u-mt-lg" variant="link" onClick={() => setFilterValue('')}>
                Clear all filters
              </Button>
            </EmptyState>
          )}
        </React.Fragment>
      ) : (
        <Bullseye className="pf-u-p-xl">
          <Spinner />
        </Bullseye>
      )}
    </div>
  </Dropdown>
);

AppFilterDropdown.propTypes = {
  isLoaded: PropTypes.bool,
  isOpen: PropTypes.bool,
  filterValue: PropTypes.string,
  setFilterValue: PropTypes.func,
  setIsOpen: PropTypes.func.isRequired,
  filteredApps: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.node.isRequired,
      routes: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string.isRequired })),
    })
  ),
};

const AppFilter = () => {
  const { filteredApps, isLoaded, isOpen, setIsOpen, filterValue, setFilterValue } = useGlobalNav();

  return (
    <React.Fragment>
      <AppFilterDropdown
        isLoaded={!!isLoaded}
        setIsOpen={setIsOpen}
        isOpen={isOpen}
        filterValue={filterValue}
        setFilterValue={setFilterValue}
        filteredApps={filteredApps}
      />
      <div className={classnames({ 'pf-c-backdrop': isOpen })} />
    </React.Fragment>
  );
};

export default AppFilter;
