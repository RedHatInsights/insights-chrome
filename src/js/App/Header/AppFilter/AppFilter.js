import React, { useEffect, useRef } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import messages from '../../../Messages';

import { CaretDownIcon } from '@patternfly/react-icons/dist/js/icons/caret-down-icon';
import FilterIcon from '@patternfly/react-icons/dist/js/icons/filter-icon';

import {
  Bullseye,
  Button,
  Dropdown,
  DropdownToggle,
  EmptyState,
  EmptyStateBody,
  EmptyStateIcon,
  EmptyStateVariant,
  Flex,
  SearchInput,
  Spinner,
  Split,
  SplitItem,
  Text,
  TextContent,
  Title,
} from '@patternfly/react-core';

import './AppFilter.scss';
import useAppFilter from './useAppFilter';
import ChromeLink from '../../Sidenav/Navigation/ChromeLink';
import { useLocation } from 'react-router-dom';

const App = ({ id, title, links = [], setIsOpen }) =>
  links.length > 0 ? (
    <div className="galleryItem">
      <Split>
        <SplitItem>
          <TextContent>
            <Text component="h4">{title}</Text>
            {links.map(({ filterable, href, title, isHidden, ...rest }) =>
              isHidden ? null : (
                <Text component="p" key={`${id}-${href}`} onClick={() => setIsOpen?.(false)}>
                  <ChromeLink {...rest} title={title} href={href}>
                    {title}
                  </ChromeLink>
                </Text>
              )
            )}
          </TextContent>
        </SplitItem>
      </Split>
    </div>
  ) : null;

App.propTypes = {
  id: PropTypes.string.isRequired,
  title: PropTypes.node,
  links: PropTypes.arrayOf(PropTypes.shape({ href: PropTypes.string.isRequired, title: PropTypes.string.isRequired })).isRequired,
  setIsOpen: PropTypes.func,
};

const AppFilterDropdown = ({ isLoaded, setIsOpen, isOpen, filterValue, setFilterValue, filteredApps }) => {
  const dropdownRef = useRef(null);
  const { pathname } = useLocation();
  const intl = useIntl();

  useEffect(() => {
    if (isLoaded) {
      setIsOpen(false);
    }
  }, [pathname]);

  return (
    <Dropdown
      className="pf-m-full-height"
      toggle={
        <DropdownToggle
          id="toggle-id"
          onToggle={(_isOpen, event) => {
            if (!dropdownRef.current?.contains(event.target)) {
              setIsOpen(!isOpen);
            }
          }}
          toggleIndicator={CaretDownIcon}
        >
          All apps and services
        </DropdownToggle>
      }
      isOpen={isOpen}
      ouiaId="App Filter"
    >
      {ReactDOM.createPortal(
        <div ref={dropdownRef} className="pf-c-dropdown chr-c-page__app-filter-dropdown-menu" data-testid="chr-c__find-app-service">
          <div className="pf-c-dropdown__menu">
            <div className="chr-app-filter-content">
              {isLoaded ? (
                <React.Fragment>
                  <Flex className="chr-l-flex-app-filter-search">
                    <SearchInput
                      data-ouia-component-id="app-filter-search"
                      placeholder={intl.formatMessage(messages.findAppOrService)}
                      value={filterValue}
                      onChange={(val) => setFilterValue(val)}
                      onClear={(e) => {
                        setFilterValue('');
                        e.stopPropagation();
                      }}
                    />
                  </Flex>
                  {filteredApps?.length > 0 ? (
                    <div className="gallery">
                      {filteredApps.map((app) => (
                        <App key={app.id} {...app} setIsOpen={setIsOpen} />
                      ))}
                    </div>
                  ) : (
                    <EmptyState className="pf-u-mt-xl" variant={EmptyStateVariant.full}>
                      <EmptyStateIcon className="pf-u-mb-xl" icon={FilterIcon} />
                      <Title headingLevel="h4">{intl.formatMessage(messages.noMatchingAppsFound)}</Title>
                      <EmptyStateBody className="pf-u-mb-xl">{intl.formatMessage(messages.tryChangingSearch)}</EmptyStateBody>
                      <Button
                        ouiaId="app-filter-clear-input"
                        className="pf-u-mt-lg"
                        variant="link"
                        onClick={(e) => {
                          setFilterValue('');
                          e.stopPropagation();
                        }}
                      >
                        {intl.formatMessage(messages.clearFilters)}
                      </Button>
                    </EmptyState>
                  )}
                </React.Fragment>
              ) : (
                <Bullseye className="pf-u-p-xl">
                  <Spinner />
                </Bullseye>
              )}
            </div>
          </div>
        </div>,
        document.body
      )}
    </Dropdown>
  );
};

AppFilterDropdown.propTypes = {
  isLoaded: PropTypes.bool,
  isOpen: PropTypes.bool,
  filterValue: PropTypes.string,
  setFilterValue: PropTypes.func,
  setIsOpen: PropTypes.func.isRequired,
  filteredApps: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.node.isRequired,
      routes: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string.isRequired })),
    })
  ),
};

const AppFilter = () => {
  const { filteredApps, isLoaded, isOpen, setIsOpen, filterValue, setFilterValue } = useAppFilter();

  return (
    <React.Fragment>
      <AppFilterDropdown
        isLoaded={isLoaded}
        setIsOpen={setIsOpen}
        isOpen={isOpen}
        filterValue={filterValue}
        setFilterValue={setFilterValue}
        filteredApps={filteredApps}
      />
    </React.Fragment>
  );
};

export default AppFilter;
