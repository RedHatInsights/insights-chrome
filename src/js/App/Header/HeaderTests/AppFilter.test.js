import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';

import AppFilter from '../AppFilter';

jest.mock('../../../nav/sourceOfTruth', () => {
  const source = jest.requireActual('../../../nav/sourceOfTruth');
  return {
    __esModule: true,
    ...source,
    default: jest.fn().mockResolvedValue({}),
  };
});
jest.mock('../../../nav/globalNav', () => {
  const source = jest.requireActual('../../../nav/globalNav');
  const appData = ['insights', 'openshift', 'cost-management', 'migrations', 'subscriptions', 'ansible', 'settings'].reduce(
    (acc, curr) => ({
      ...acc,
      [curr]: {
        id: curr,
        title: curr,
        routes: [{ id: curr, title: curr }],
      },
    }),
    {}
  );
  return {
    __esModule: true,
    ...source,
    getNavFromConfig: jest.fn().mockImplementation(() => Promise.resolve(appData)),
  };
});

describe('<AppFilter />', () => {
  test('should render correctly', () => {
    const { container } = render(<AppFilter />);
    expect(container).toMatchSnapshot();
  });

  test('should open and fetch data', async () => {
    const { container } = render(<AppFilter />);
    const button = container.querySelector('#toggle-id');
    await act(async () => {
      fireEvent.click(button);
    });
    expect(container.querySelectorAll('.content')).toHaveLength(1);
  });

  test('should set and clear filter input value', async () => {
    const { container } = render(<AppFilter />);
    const button = container.querySelector('#toggle-id');
    await act(async () => {
      fireEvent.click(button);
    });
    const input = container.querySelector('input');
    await act(async () => {
      fireEvent.change(input, { target: { value: 'foo' } });
    });
    expect(input.value).toEqual('foo');
    const clearButton = container.querySelector('[data-ouia-component-id=app-filter-search] button');
    expect(clearButton).toBeTruthy();
    await act(async () => {
      fireEvent.click(clearButton);
    });
    expect(input.value).toEqual('');
  });

  test('should render empty state on no filter match and clear app filters', async () => {
    const { container } = render(<AppFilter />);
    const button = container.querySelector('#toggle-id');
    await act(async () => {
      fireEvent.click(button);
    });
    const input = container.querySelector('input');
    await act(async () => {
      fireEvent.change(input, { target: { value: 'foo' } });
    });
    expect(container).toMatchSnapshot();
    const clearButton = container.querySelector('[data-ouia-component-id=app-filter-clear-input]');
    expect(clearButton).toBeTruthy();
    await act(async () => {
      fireEvent.click(clearButton);
    });
    expect(input.value).toEqual('');
  });
});
