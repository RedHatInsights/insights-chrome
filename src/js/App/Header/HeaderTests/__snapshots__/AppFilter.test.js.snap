// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AppFilter /> should render correctly 1`] = `
<div>
  <div
    class="pf-c-dropdown ins-c-page__app-filter-dropdown"
    data-ouia-component-id="App Filter"
    data-ouia-component-type="PF4/Dropdown"
    data-ouia-safe="true"
  >
    <button
      aria-expanded="false"
      aria-haspopup="false"
      class="pf-c-dropdown__toggle"
      data-ouia-component-id="OUIA-Generated-DropdownToggle-1"
      data-ouia-component-type="PF4/DropdownToggle"
      data-ouia-safe="true"
      id="toggle-id"
      type="button"
    >
      <span
        class="pf-c-dropdown__toggle-text"
      >
        Find an app or service
      </span>
      <span
        class="pf-c-dropdown__toggle-icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 320 512"
          width="1em"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    class=""
  />
</div>
`;

exports[`<AppFilter /> should render empty state on no filter match and clear app filters 1`] = `
<div>
  <div
    class="pf-c-dropdown pf-m-expanded ins-c-page__app-filter-dropdown"
    data-ouia-component-id="App Filter"
    data-ouia-component-type="PF4/Dropdown"
    data-ouia-safe="true"
  >
    <button
      aria-expanded="true"
      aria-haspopup="false"
      class="pf-c-dropdown__toggle"
      data-ouia-component-id="OUIA-Generated-DropdownToggle-4"
      data-ouia-component-type="PF4/DropdownToggle"
      data-ouia-safe="true"
      id="toggle-id"
      type="button"
    >
      <span
        class="pf-c-dropdown__toggle-text"
      >
        Find an app or service
      </span>
      <span
        class="pf-c-dropdown__toggle-icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 320 512"
          width="1em"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          />
        </svg>
      </span>
    </button>
    <div
      class="pf-c-dropdown__menu"
    >
      <div
        class="content"
      >
        <div
          class="pf-l-flex search"
        >
          <div
            class="pf-c-search-input"
            data-ouia-component-id="app-filter-search"
          >
            <span
              class="pf-c-search-input__text"
            >
              <span
                class="pf-c-search-input__icon"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                  />
                </svg>
              </span>
              <input
                aria-label="Search input"
                class="pf-c-search-input__text-input"
                placeholder="Find application or service"
                value="foo"
              />
            </span>
            <span
              class="pf-c-search-input__utilities"
            >
              <span
                class="pf-c-search-input__clear"
              >
                <button
                  aria-disabled="false"
                  aria-label="Clear"
                  class="pf-c-button pf-m-plain"
                  data-ouia-component-id="OUIA-Generated-Button-plain-2"
                  data-ouia-component-type="PF4/Button"
                  data-ouia-safe="true"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    fill="currentColor"
                    height="1em"
                    role="img"
                    style="vertical-align: -0.125em;"
                    viewBox="0 0 352 512"
                    width="1em"
                  >
                    <path
                      d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
                    />
                  </svg>
                </button>
              </span>
            </span>
          </div>
        </div>
        <div
          class="pf-c-empty-state pf-u-mt-xl"
        >
          <div
            class="pf-c-empty-state__content"
          >
            <svg
              aria-hidden="true"
              class="pf-c-empty-state__icon pf-u-mb-xl"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
              />
            </svg>
            <h4
              class="pf-c-title pf-m-md"
            >
              No matching applications or services found.
            </h4>
            <div
              class="pf-c-empty-state__body pf-u-mb-xl"
            >
              This filter criteria matches no applications or services. Try changing your input filter.
            </div>
            <button
              aria-disabled="false"
              class="pf-c-button pf-m-link pf-u-mt-lg"
              data-ouia-component-id="app-filter-clear-input"
              data-ouia-component-type="PF4/Button"
              data-ouia-safe="true"
              type="button"
            >
              Clear all filters
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="pf-c-backdrop"
  />
</div>
`;
